teps/0060-remote-resource-resolution.md

```diff
@@ -242,207 +255,562 @@ users can leverage those shared pipelines.
+  resource:
+    image_url: gcr.io/tekton-releases/catalog/upstream/golang-fuzz:0.1
+    name: golang-fuzz
+    signer: cosign # TBD. See TEP-0091.
+```
+
+```yaml
+taskRef:
+  resolver: git
+  resource:
+    repository_url: https://github.com/tektoncd/catalog.git
+    branch: main
+    path: /task/golang-fuzz/0.1/golang-fuzz.yaml
+```
+
+`pipelineRef` new syntax samples:
+
+```yaml
+pipelineRef:
+  resolver: git
+  resource:
+    repository_url: https://github.com/sbwsg/experimental.git
+    branch: remote-resolution
+    path: /remote-resolution/pipeline.yaml
+```
+
+```yaml
+pipelineRef:
+  resolver: in-cluster
+  resource:
+    name: my-team-build-pipeline
+```
+
+Note that the fields under `resource:` are validated by a resolver, not
+by Tekton Pipelines. [TEP-0075 (Dictionary
+Params)](https://github.com/tektoncd/community/pull/479) describes
+adding support for JSON object schema to Params. It would make sense to
+bring in the same schema support for the `resource:` field as well and
+eventually resolvers might be able to publish the schema they accept so
+that Pipelines can enforce it during validation of a `TaskRun` or
+`PipelineRun`.
+
+### 2. Implement procedure for resolution: `ResourceRequest` CRD
+
+When the Pipelines reconciler sees a `taskRef` or `pipelineRef` with a
+`resolver` field it creates a `ResourceRequest` object. The object is
+populated using fields from the `taskRef` or `pipelineRef`.
+
+Once resolved the contents of the requested resource are base64-encoded
+and stored in-line in the `status.data` field of the `ResourceRequest`.
+Metadata is set in the `status.annotations` field. Examples of metadata
+include the commit digest of a resolved `git` resource, a bundle's
+digest, or the resolved data's `content-type`. The `ResourceRequest` is
+marked as successfully resolved. Here is a sample of a resolved
+`ResourceRequest` with some metadata fields trimmed out for brevity:
+
+```yaml
+apiVersion: resolution.tekton.dev/v1alpha1
+kind: ResourceRequest
+spec:
+  params:
+    repository_url: https://github.com/sbwsg/experimental.git
+    branch: remote-resolution
+    path: /remote-resolution/pipeline.yaml
+status:
+  data: a2luZDogUGlwZWxpbmUKYXBpVmVyc2lvbjogdGVrdG9uLmRldi92MWJldGExCm1ldGFkYXRhOgogIG5hbWU6IHAKc3BlYzoKICB0YXNrczoKICAtIG5hbWU6IGZldGNoLWZyb20tZ2l0CiAgICB0YXNrU3BlYzoKICAgICAgc3RlcHM6CiAgICAgIC0gaW1hZ2U6IGFscGluZS9naXQKICAgICAgICBzY3JpcHQ6IHwKICAgICAgICAgIGdpdCBjbG9uZSBodHRwczovL2dpdGh1Yi5jb20vdGVrdG9uY2QvcmVzdWx0cwo=
+  annotations:
+    commit: 2c9b093e94f30f588dc798cc56a2559d4da9d573
+    content-type: application/x-yaml
+  conditions:
+  - lastTransitionTime: "2021-10-14T15:38:14Z"
+    status: "True"
+    type: Succeeded
+```
+
+If a `ResourceRequest` fails resolution then it is marked as failed with
+a `Succeeded/"False"` condition. The condition will include a
+machine-readable `reason` and human-readable `message` explaining the
+failure.
+
+### 3. Create a new Tekton Resolution project
```

Comment:
Is this now going to be a non-experimental project? If so, is the intention to keep this out of `tektoncd/pipeline` longterm?
I think it's fine to experiment with this a bit, but eventually I would see this living in `tektoncd/pipeline` directly.

Reply:
Ah - good catch, this is really intended to be a production-ready project, definitely not experimental anymore, with its own project in the `tektoncd` org. I'll update the TEP with more detail, cheers!

I'm a bit unsure how to proceed wrt this living in `tektoncd/pipeline`. Would it be possible to clarify in more precise terms what you want to live in the Pipelines project? The reason I request this is that previous review cycles on this TEP (and in discussions before it was created) have really pushed back on that idea a lot, so I've intentionally directed the design away from it over the past six months. Here are a number of instances of feedback I am thinking about:

- ["To that end, perhaps the one missing requirement and use case from the provided list is that the remote resolution process should no longer be embedded into Tekton. Instead, the goal is to offer Tekton operators and by extension users, a pluggable framework for resolving tasks and pipelines with two reference implementations, local and Tekton Bundles."](https://github.com/tektoncd/community/pull/389#issuecomment-809738284)
- ["...in triggers for example, we refer to a set of interceptor. The mechanism used in tektoncd/pipeline could be the same/re-used in triggers."](https://github.com/tektoncd/community/pull/389#discussion_r629256004)
- [A very detailed comment from @vdemeester about trade-offs and benefits of implementing as a new Tekton component.](https://github.com/tektoncd/community/pull/389#pullrequestreview-655730552)
- The conversation thread in [pipelines issue 2298 (Add support for referencing Tasks in git)](https://github.com/tektoncd/pipeline/issues/2298) that originally started a lot of this design work. The core of the question related to "where" this logic should live starts [with this comment](https://github.com/tektoncd/pipeline/issues/2298#issuecomment-721233501) I think.

I guess my question becomes: how do we _resolve_ this design question once and for all? (...pun intended). Should this be a separate project or should we bake this functionality directly into Tekton Pipelines?

Reply:
I've updated the TEP with the production-readiness note about this new project.

Reply:
> I guess my question becomes: how do we resolve this design question once and for all? (...pun intended). Should this be a separate project or should we bake this functionality directly into Tekton Pipelines?

My usual answer here is : if we can do it in a separate project let's do it, and see in the future if it make sense to bake it into pipeline. The more tektoncd/pipeline mature / gets adopted, the more causious we have to be into getting things in 😅

Reply:
I also tend to like the idea of keeping the responsibilities separated along project boundaries here, for a couple of reasons:

- Faster iteration times bootstrapping a new project vs dumping giant PRs on the Pipelines team.
- Reduce the chance of affecting other parts of Pipeline's codebase.
- The whole "yes is forever" thing. We can always revisit this decision.

I'll raise this in the API WG when I'm back from holiday (Nov 29). I am hopeful we can figure out how to proceed on this one with just a bit more discussion.

Reply:
Feedback from the API WG just now: will this be embedded in Tekton Pipelines directly eventually? Maybe - might make sense or might not. Tekton Operator can be configured to deploy this, Pipelines itself could embed it, or it could be separate forever. In the short term keeping it separate would work, or placing it in Pipelines and extracting it out later if we decide it's a bad fit might work as well.

@afrittoli  - do you have any thoughts what you'd like to see here to close this discussion item? I think this is the last open point that needs to be resolved and I'd like to keep the TEP moving if possible.

Reply:
my 2 cents, either option seems fine for now, doesnt seem like it would be too hard to change later

- separate project: cons: more overhead now (creating, releasing and maintaining a separate project, not clear how to bundle this easily for users
- inside tekton pipelines: cons: moving into a separate project later = impacts folks who rely on it (but if ppl are relying on it, maybe we wouldnt want to move it out?